src/arkode/arkode_mristep.c

```diff
@@ -1428,35 +1658,91 @@ int mriStep_FullRHS(ARKodeMem ark_mem, sunrealtype t, N_Vector y, N_Vector f,
            <0 => step encountered unrecoverable failure
   ---------------------------------------------------------------*/
-int mriStep_TakeStep(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+int mriStep_TakeStepMRIGARK(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
 {
-  ARKodeMRIStepMem step_mem; /* outer stepper memory       */
-  int is;                    /* current stage index        */
-  int retval;                /* reusable return flag       */
+  ARKodeMRIStepMem step_mem;          /* outer stepper memory       */
+  int is;                             /* current stage index        */
+  int retval;                         /* reusable return flag       */
+  N_Vector tmp;                       /* N_Vector pointer           */
+  SUNAdaptController_Type adapt_type; /* timestep adaptivity type   */
+  sunrealtype t0, tf;                 /* start/end of each stage    */
+  sunbooleantype calc_fslow;
+  sunbooleantype need_inner_dsm;
+  sunbooleantype do_embedding;
+  int nvec;
+
+  /* access the MRIStep mem structure */
+  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
 
   /* initialize algebraic solver convergence flag to success;
      error estimate to zero */
   *nflagPtr = ARK_SUCCESS;
   *dsmPtr   = ZERO;
 
-  /* access the MRIStep mem structure */
-  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
-  if (retval != ARK_SUCCESS) { return (retval); }
+  /* initial time and state for step */
+  N_VScale(ONE, ark_mem->yn, ark_mem->ycur);
+  t0 = ark_mem->tn;
+
+  /* determine whether embedding stage is needed */
+  do_embedding = !ark_mem->fixedstep ||
+                 (ark_mem->AccumErrorType != ARK_ACCUMERROR_NONE);
+
+  /* if MRI adaptivity is enabled: reset fast accumulated error,
+     and send appropriate control parameter to the fast integrator */
+  adapt_type     = SUNAdaptController_GetType(ark_mem->hadapt_mem->hcontroller);
+  need_inner_dsm = SUNFALSE;
+  if (adapt_type == SUN_ADAPTCONTROLLER_MRI_TOL)
+  {
+    need_inner_dsm      = SUNTRUE;
+    step_mem->inner_dsm = ZERO;
+    retval              = mriStepInnerStepper_ResetError(step_mem->stepper);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to reset the inner stepper error estimate");
+      return (ARK_INNERSTEP_FAIL);
+    }
+    retval =
+      mriStepInnerStepper_SetRTol(step_mem->stepper,
+                                  step_mem->inner_control * ark_mem->reltol);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to set the inner stepper tolerance");
+      return (ARK_INNERSTEP_FAIL);
+    }
+  }
+
+  /* if any temporal adaptivity is enabled: reset the inner integrator
+     to the beginning of this step (in case of recomputation) */
```

Comment:
I do not recall seeing any MRI-GARK method where the embedding was handled via a simple ERK/DIRK update, as opposed to a fast integration.  I like the idea of doing the embedding out of order (that's what I do with MERK methods).  I'll look into that now, but if it will take too long then we could defer it for later.

Reply:
While doing the embedding out of order could allow us to skip this `Reset`, I am unsure how to know whether this step attempt immediately follows a rejected vs a successful step.  I suppose that I could add in a counter to the MRIStep memory structure to try to keep track of this (like @maggul did for LSRKStep), but it would be nice to know if there's a simpler approach.

Reply:
> While doing the embedding out of order could allow us to skip this `Reset`, I am unsure how to know whether this step attempt immediately follows a rejected vs a successful step. I suppose that I could add in a counter to the MRIStep memory structure to try to keep track of this (like @maggul did for LSRKStep), but it would be nice to know if there's a simpler approach.

Nevermind, `nflagPtr` stores the value `PREV_ERR_FAIL` on the step immediately following a rejected step attempt, so I can use that.